lib/resumption/resumable.go

```diff
@@ -0,0 +1,334 @@
+// Teleport
+// Copyright (C) 2023  Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+func newResumableConn(localAddr, remoteAddr net.Addr) *Conn {
+	r := &Conn{
+		managedConn: managedConn{
+			localAddr:  localAddr,
+			remoteAddr: remoteAddr,
+		},
+	}
+	r.cond.L = &r.mu
+	return r
+}
+
+// Conn is a [net.Conn] whose underlying transport can be closed and reopened,
+// to maintain the illusion of a perfect unbroken stream of bytes even if
+// network conditions would otherwise terminate a normal connection.
+type Conn struct {
+	managedConn
+
+	// attached is non-nil iff there's an underlying connection attached;
+	// calling it should eventually result in the connection becoming detached,
+	// signaled by the field becoming nil.
+	attached func()
```

Comment:
nit: might a better name be `detachUnderlying`?

Reply:
Renamed to `requestDetach`.